extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -2197,6 +2199,19 @@ private Pair<ControllerQueryKernel, ListenableFuture<?>> run() throws IOExceptio
       startTaskLauncher();
 
       boolean runAgain;
+      final QueryContext queryContext = querySpec.getContext();
+
+      // Fetch the timeout, but don't use default server configured timeout if the user has not specified one.
```

Comment:
For cleanliness, would be good to break the retrieval of `queryFailDeadline` out into its own `private` method. You can set the deadline to `DateTimes.MAX` if there's no timeout, so there isn't a need to also have `hasTimeout`.